Markup and style to Events index page

marioabreu opened the pull request.

marioabreu committed:
Markup and style to Events index page

azevedo commented on app/assets/stylesheets/components/_body.sass:
```diff
@@ -0,0 +1,4 @@
+$Body-color-background: Theme-color(background, body)
```
let's go with the camel case convention for expressions. $Body-backgroundColor

azevedo-252 commented:
I've stopped reviewing when I saw hex colors in your Button component. Please review your code before creating the PR. This isn't me being a pain in the a**, it's just a matter of not "wasting" the reviewer's time, and make the most of the actual review, i.e. a code review, ultimately, isn't supposed to be fixing the commiter's mistakes, it's just a second pair of eyes for giving insights regarding implementation details, the feature as whole, ....
Of course, in your case, it's different because your learning, but there are a few things that you can already do by yourself.
So keep up the good the work, and try to remember this.

> Keep calm... and commit wisely

marioabreu committed:
Reviewed some errors of the previews commit.
Also, I have changed the some index's keys of the background and text,
in order to try to be decaupled from the HTML elements.

azevedo commented on app/assets/stylesheets/components/_body.sass:
```diff
@@ -0,0 +1,4 @@
+$Body-color-background: Theme-color(background, default)
```
I would go with `$Body-backgroundColor`

marioabreu replied:
:+1:

azevedo commented on app/assets/stylesheets/components/_button.sass:
```diff
@@ -0,0 +1,19 @@
+$Button-default-backgroundColor: Theme-color(button, default)
```
Please let's try to follow the left to right (generic to specific) convention

azevedo replied:
Have a look at all partials when you do this

azevedo commented on app/assets/stylesheets/components/_button.sass:
```diff
@@ -0,0 +1,19 @@
+  border: 1px solid darken($Button-default-backgroundColor, 10%)
+  color: $Button-text-color
+  line-height: $Button-height
+  margin-top: $Button-spacing
```
> Most components should not set their own width, margin, and positioning.

azevedo commented on app/assets/stylesheets/components/_button.sass:
```diff
@@ -0,0 +1,19 @@
+  color: $Button-text-color
+  line-height: $Button-height
+  margin-top: $Button-spacing
+  padding: 10px 20px 10px 20px
```
1. You could write: `padding: 10px 20px`
2. Where do these values come from?

marioabreu replied:
I was making the button, and this values seems nice to me. But yeah, it does not follow any method.
I changed to use the `padding: $Theme-spacing-small $Theme-spacing-small`, sounds good?

azevedo commented on app/assets/stylesheets/components/_button.sass:
```diff
@@ -0,0 +1,19 @@
+$Button-height: 50px
+$Button-spacing: $Theme-spacing-base
+
+.Button
```
Explain to me how does this component work?
There's the component `Button` and then there's a descendant `Button-default`?
Why?
What's the reason for having a descendant `Button-default`?

marioabreu replied:
Yeah, it doesn't make sense... 
Although my idea was to have a button with no colors, only spacing border, and text alignment. Maybe this is a perfect example to use modifiers, but one thing i can't understand is where do i set the actual size of the button?

azevedo commented on app/assets/stylesheets/components/_container.sass:
```diff
@@ -0,0 +1,5 @@
+$Container-maxWidth: 960px
+
+.Container
+  max-width: $Container-maxWidth
```
Property order

azevedo commented on app/assets/stylesheets/components/_event.sass:
```diff
@@ -0,0 +1,30 @@
+$Event-color-background: Theme-color(background, default)
```
This name is too generic. It's always better when you avoid generic names.

marioabreu replied:
I changed it to `Event-backgroundColor` as you said in the body component. In this event component, how can i be specific without be coupled to the HTML? at the time, The event is a box and the background color for the content. Would `Event-backgroundColor-content` be a good name?

azevedo commented on app/assets/stylesheets/components/_event.sass:
```diff
@@ -0,0 +1,30 @@
+  background-color: $Event-color-background
+  border-radius: 3px
+  border-top: 5px solid $Event-color-borderTop
+  box-shadow: 0 2px 10px -3px gray
```
Why is a CSS color being used here?

marioabreu replied:
Escaped me.
I will put it on the `_theme.scss` under the map `background:( shadow: $Theme-gray)`. It make sense?
Then i can call it with `$Event-color-shadow: Theme-color(background, shadow)`

azevedo replied:
I'm still not happy with the fact that you're inserting every background you think of into that bag. But whenever you have your final design language it will be easier to get to something better

azevedo commented on app/assets/stylesheets/components/_event.sass:
```diff
@@ -0,0 +1,30 @@
+  box-shadow: 0 2px 10px -3px gray
+  display: inline-block
+  height: $Event-height
+  margin: $Event-spacing-base
```
Again, the component shouldn't set its own margin.
You must think about the component in isolation of everything. That margin only exists because, in the scenario you're using it, when there are Events side by side, you need them to be separated. Well that's not this component's responsibility

marioabreu replied:
I can't came up with a solution for this to work. I know that is the ancestor that is responsible to the positioning, but can't work it out, because for this case I have two kinds of arrangement. (aligned to right on events list and on center for the button to create a new button.)

azevedo commented on app/assets/stylesheets/components/_event.sass:
```diff
@@ -0,0 +1,30 @@
+    background-color: lighten($Event-color-borderTop, 30%)
+
+.Event-info
+  list-style: none
```
You could remove this in the Theme component, directly on the _li_ element

marioabreu replied:
:+1:

azevedo commented on app/assets/stylesheets/components/_event.sass:
```diff
@@ -0,0 +1,30 @@
+  color: $Event-text-title-color
+  text-decoration: none
+
+.Event-create
```
I can understand `Event-title`, `Event-info`, but this one isn't so obvious. Why is it called `create`?

marioabreu replied:
I don't think this is a good name, my idea was to have a separation from the events list, which is displaying the existing events from the left to the right. After that list, i think of place a button at the bottom. Yeah, but maybe it doesn't make sense for this to belong to this event, maybe on Body element.

azevedo commented on app/assets/stylesheets/components/_header.sass:
```diff
@@ -1,9 +1,9 @@
 $Header-spacing-base: $Theme-spacing-base
 $Header-height-header: 100px
-$Header-background-color: $Theme-white
+$Header-color-background: $Theme-white
```
I see what you're doing here. But wouldn't you agree, that background color should be seen as a whole world, i.e. `$Header-backgroundColor`?
What do you think?

marioabreu replied:
:+1:

azevedo commented on app/assets/stylesheets/components/_theme.scss:
```diff
@@ -14,31 +14,28 @@ $Theme-fontSizes: (
 $Theme-color: (
   background: (
-    body: $Theme-white,
+    default: $Theme-white,
```
What made you choose this name and merge those 3 others?

marioabreu replied:
It was mainly for two reasons. First, as you said before, these names have a strong connection to the HTML (header, body and footer). The second reason was because it was because the values were the same, `$Theme-white` for all of them. Maybe part of the problem was caused because there are a color palette well defined.

azevedo commented on app/assets/stylesheets/components/_theme.scss:
```diff
@@ -14,31 +14,28 @@ $Theme-fontSizes: (
   text: (
-    body: $Theme-blue,
+    default: $Theme-blue,
+    lighter: $Theme-white,
```
`Lighter` has the exact same problem as coupling to a color name

marioabreu replied:
:+1: 
I agree. at the time it make more sense to me. What do you think of a `highlight: $Theme-white` or `title: $Theme-white` ? This is to be used on dark background.

azevedo commented on app/views/events/index.slim:
```diff
@@ -1,11 +1,13 @@
-      li.fa.fa-user = "#{event.attendees.size}"
+  .Event
+    = link_to event.name, event_path(event.id), class: 'Event-title'
+    ul
```
Why is this a _list_?

azevedo replied:
Perhaps you wanted to do something like:

```
ul
  -@events.each do |event|
    = link_to event.name, event_path(event.id), class: 'Event-title'
      li.Event
        .Event-info
          b = "On #{Date::DAYNAMES[event.start_date.wday]}"
        .Event-info = event.starting_date
        .Event-info = event.ending_date
          .fa.fa-user = "#{event.attendees.size}"
```

marioabreu replied:
:+1:

azevedo commented on app/views/events/index.slim:
```diff
@@ -1,11 +1,13 @@
+    = link_to event.name, event_path(event.id), class: 'Event-title'
+    ul
+      li.Event-info
+        b = "On #{Date::DAYNAMES[event.start_date.wday]}"
```
What's the `b`. Currently it's only used for styling. So if that's the case here, you should do it with CSS. If it's for emphasis you have `em`

marioabreu replied:
Yes, it was for emphasis.

azevedo replied:
This is still wrong

azevedo commented on app/views/events/index.slim:
```diff
@@ -1,11 +1,13 @@
+        li.fa.fa-user = "#{event.attendees.size}"
 -if !current_user.nil?
-  = link_to 'New Event', new_event_path
+  .Event-create
```
I don't get where this div should be? And to what component it belongs to. The name indicates it belongs to `Event`, but the markup doesn't agree with it

azevedo commented on app/views/events/index.slim:
```diff
@@ -1,11 +1,13 @@
 -if !current_user.nil?
-  = link_to 'New Event', new_event_path
+  .Event-create
+    .Button
```
Again, I don't understand why there's a `Button` and a `Button-default`.
It simply could be `= link_to 'New Event', new_event_path, class: 'Button'`
And if it comes to that create a specific modifier like `Button--default`.

azevedo commented on app/views/layouts/application.slim:
```diff
@@ -12,30 +12,33 @@ html lang="en"
+            .Logo
+              = link_to "Afterwork", root_path, class: 'Logo-link'
+          .Header-nav
+            nav.Navbar role="navigation"
```
Why is `.Navbar` attached to `nav`? I'm looking at your markup and it seems to make more sense if it was attached to the `ul`

marioabreu replied:
:+1:

marioabreu committed:
Refactored as suggested.

marioabreu committed:
Refactor as suggested.

I have made this changes in other branch, however was not a good idea.

marioabreu committed:
added flex-shrink property to MainLayout-notificiation

marioabreu committed:
Corrected as suggested.
Deleted end_date from Events.

marioabreu committed:
Removed the extra li.
Correct the margins of the grid.

marioabreu committed:
Correct the margins of the Event boxes.

azevedo commented on app/assets/stylesheets/components/_event.sass:
```diff
@@ -0,0 +1,24 @@
+$Event-backgroundColor: Theme-color(background, default)
+$Event-borderTop-color: #FF623F  // FIXME Put this color on background color MAP on theme.sass file.
```
Perhaps `$Event-color-borderTop` ?

This makes me wonder about `$Event-backgroundColor`. Do you think `$Event-color-background` would make more sense?

marioabreu replied:
Yes, you are right. From generic to specific, left to right...
It could be. only the backgroundColor is the css property name, maybe it make more sense to have it `Event-backgroundColor`. Or we can have always the template `Component-color-terget`. Either way, it must be consistent. Should we go with the second, `Event-color-background`?

azevedo replied:
I think we can go with the second one yes. It makes more sense to me at the moment :)

marioabreu replied:
:+1:

azevedo commented on app/assets/stylesheets/components/_grid.sass:
```diff
@@ -0,0 +1,12 @@
+.Grid
+  margin-bottom: $Grid-margin-base
+  margin-left: -$Grid-margin-base
+  margin-top: $Grid-margin-base
```
The same issue I raised regarding margin-left applies here as well. Please inspect what this produces and confirm it is what you want.
You either make it negative, which is an internal style rather than an external (which would go against "**Most components should not set their own width, margin, and positioning**"), or simply remove this.

marioabreu replied:
I will remove it.

azevedo commented on app/views/events/index.slim:
```diff
@@ -1,11 +1,13 @@
-  = link_to 'New Event', new_event_path
+ul.Grid
+  -@events.each do |event|
+    .Grid-item
```
I think it would make more sense, semantically, for the `li` to be attached to `Grid-item` instead of `Event`.
What do you think?

marioabreu replied:
Yes, it makes more sense.

azevedo commented on app/views/events/index.slim:
```diff
@@ -1,11 +1,13 @@
+          .Event-title = event.name
+          .Event-info
+            em = "On #{Date::DAYNAMES[event.start_date.wday]}"
+          .Event-info = event.starting_date
```
These two lines are confusing.

```
.Event-info = event.starting_date
  fa.fa-user = "#{event.attendees.size}"
```

What's the html generated?

marioabreu replied:
The second line wasn't generating anything.
The idea was to present a user icon followed by the number of attendees.
I came up with 

```
.Event-info 
  span.fa.fa-user = "#{event.attendees.size}"
```

azevedo commented on app/views/events/index.slim:
```diff
@@ -1,11 +1,13 @@
+          .Event-info = event.starting_date
+            fa.fa-user = "#{event.attendees.size}"
+.Body-action
+  -if !current_user.nil?
```
`- if current_user.present?`

marioabreu replied:
:+1:

marioabreu committed:
Changed naming from backgroundColor to color-brackground in components.
Removed the top margin from the grid component.
Change the !current_user.nil? to current_user.present?, ir is more readable.

azevedo-252 commented:
:+1:

azevedo-252 commented:
I hope this PR will serve as a knowledge center. A lot of common mistakes have been covered here, as well as some interesting discussions.
